datapath: remove datapath abstraction / interface

mhofstetter opened the pull request:
Previously, the datapath abstraction was used to access datapath related components via a single interface.

With the progressing modularization of the Cilium codebase, the need for this abstraction became obsolete.

Therefore, this PR removes the abstraction. Existing code that makes use of the datapath abstraction are refactored to use Hives dependency mechanism to access the datapath related components directly.

Please review the individual commits.

Note: IMO, if we want to come up with new "coarse-grained" abstractions provided as Hive Cells in the future, we should force other modules to use the abstraction by providing the "inner" components as `ProvidePrivate` only.

mhofstetter committed:
loader: remove unused functions from IPTablesManager

This commit removes some unused functions from the interface `IPTablesManager`.

* `InstallProxyRules(proxyPort uint16, name string)`
* `SupportsOriginalSourceAddr() bool`
* `GetProxyPorts() map[string]uint16`

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
nodehandler: move node ID restoration & manager subscription

Currently, nodehandler related lifecycle events are registered outside
the nodehandler cell.

* node id restoration is registered in the global datapath cell
* node manager subscription happens in the daemon setup

To prepare removing the NodeHandler from the Datapath interface, this
commit moves these lifecycle event registrations into the cell of the
nodehandler itself.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
daemon: explicit dep on NodeHandler to init IPSec key watcher

Currently, initializing the IPSec key watcher during the daemon
initialization requires accessing the NodeHandler via the datapath
abstraction.

As preparation to remove the NodeHandler from the datapath abstraction,
this commit introduces an explicit dependency from the daemon to the
NodeHandler.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
datapath: Separate NodeHandler from Datapath interface

Previously, the NodeHandler was part of the Datapath interface. Since we
are no longer using the datapath interface to abstract the NodeHandler,
it seems logical to not provide the NodeHandler via the datapath
interface.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
datapath: Separate NodeNeighbors from Datapath interface

Previously, the NodeNeighbors was part of the Datapath interface. Since we
are no longer using the datapath interface to abstract the NodeNeighbors,
it seems logical to not provide it via the datapath
interface.

Existing usages of the NodeNeighbors via the Datapath interface are refactored
to have an explicit dependency to the NodeNeighbors.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
datapath: Separate WireguardAgent from Datapath interface

Previously, the WireguardAgent was part of the Datapath interface. Since we
are no longer using the datapath interface to abstract the WireguardAgent,
it seems logical to not provide it via the datapath interface.

Existing usages of the WireguardAgent via the Datapath interface are refactored
to have an explicit dependency to it.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
datapath: remove unused dependencies from datapath cell

This commit removes unused dependencies from the datapath cell.

* `NodeManager`
* `TunnelConfig`
* `Loader`
* `ModulesManager`
* `NodeMap`
* `Devices`
* `StateDB`

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
datapath: remove unused field MTU from datapathparams

This commit removes the unused field `MTU` from the struct
`DatapathParams`.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
datapath: Separate ConfigWriter from Datapath interface

Previously, the ConfigWriter was part of the Datapath interface. Since we
are no longer using the datapath interface to abstract the ConfigWriter,
it seems logical to not provide it via the datapath interface.

Existing usages (only tests) of the ConfigWriter are refactored accordingly.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
datapath: Separate NodeAdressing from Datapath interface

Previously, the NodeAdressing was part of the Datapath interface. Since we
are no longer using the datapath interface to abstract the NodeAdressing,
it seems logical to not provide it via the datapath interface.

Existing usages of the NodeAdressing via the Datapath interface (CTMap GC)
are refactored to have an explicit dependency to it.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
datapath: Separate LBMap from Datapath interface

Previously, the LBMap was part of the Datapath interface. Since we
are no longer using the datapath interface to abstract the LBMap,
it seems logical to not provide it via the datapath interface.

Existing usages of the LBMap via the Datapath interface are refactored
to have an explicit dependency to it.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
datapath: Separate Orchestrator from Datapath interface

Previously, the Orchestrator was part of the Datapath interface. Since we
are no longer using the datapath interface to abstract the Orchestrator,
it seems logical to not provide it via the datapath interface.

The only existing usages of the Orchestrator via the Datapath interface
are refactored to have an explicit dependency to it and the
`regeneration.Owner` now provides direct access to the Orchestrator.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
datapath: Separate BandwidthManager from Datapath interface

Previously, the BandwidthManager was part of the Datapath interface. Since we
are no longer using the datapath interface to abstract the BandwidthManager,
it seems logical to not provide it via the datapath interface.

The only existing usages of the BandwidthManager via the Datapath interface
are refactored to have an explicit dependency to it and the
`regeneration.Owner` now provides direct access to the BandwidthManager.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
datapath: Separate IPTablesManager from Datapath interface

Previously, the IPTablesManager was part of the Datapath interface. Since we
are no longer using the datapath interface to abstract the IPTablesManager,
it seems logical to not provide it via the datapath interface.

The only existing usages of the IPTablesManager via the Datapath interface
are refactored to have an explicit dependency to it and the
`regeneration.Owner` now provides direct access to the IPTablesManager.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
datapath: remove datapath abstraction

The datapath abstraction no longer provides access to datapath
related components.

Therefore, this commit removes the interface `types.Datapath` and
struct `linuxDatapath` - including all the referencing methods.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter commented on pkg/datapath/cells.go:
```diff
@@ -172,77 +180,26 @@ func newWireguardAgent(lc cell.Lifecycle, sysctl sysctl.Sysctl) *wg.Agent {
-	WgAgent *wg.Agent
+	LC cell.Lifecycle
 
 	// Force map initialisation before loader. You should not use these otherwise.
```
should we move this `BpfMaps` dependency to the loader cell?

joamaki replied:
Yes it should definitely move there!

mhofstetter replied:
ok, i moved it

joamaki commented on pkg/datapath/linux/node.go:
```diff
@@ -112,6 +114,19 @@ func NewNodeHandler(
+	lifecycle.Append(cell.Hook{
+		OnStart: func(_ cell.HookContext) error {
+			handler.RestoreNodeIDs()
+			nodeManager.Subscribe(handler)
```
I would consider doing the `Subscribe` already inside `NewNodeHandler`. This way the subscription is set up before anything has started.

mhofstetter replied:
Thanks, good point -> refactored.

(At first i thought that this might results in errors during `RestoreNodeIDs` checking for already existing node ids. (i faced this one before moving the subscription from the daemon here too) But this shouldn't occur in this case as the subscription shouldn't add any nodes before startup ))

joamaki commented on daemon/cmd/daemon_main.go:
```diff
@@ -1662,6 +1662,7 @@ type daemonParams struct {
 	K8sResourceSynced      *k8sSynced.Resources
 	K8sAPIGroups           *k8sSynced.APIGroups
 	NodeManager            nodeManager.NodeManager
+	NodeHandler            datapath.NodeHandler
```
Hmm, just realized that `NodeHandler` is implemented by multiple components, but here we're using it to explicitly mean the "linuxNodeHandler". This is super confusing. WDYT about adding a specific type for this? Maybe just `type LinuxNodeHandler datapath.NodeHandler` or `type DatapathNodeHandler datapath.NodeHandler`?

Could also make sense to move the `NodeHandler` interface into `pkg/node/manager`? Or I suppose alternatively we could have `datapath.NodeHandler` (implemented by linuxNodeHandler) and separately `manager.NodeHandler` that is relevant subset of `datapath.NodeHandler` that `NodeManager` invokes...

mhofstetter replied:
Good point! But if possible, I'd like to refactor this in a separate PR. IMO `datapath.NodeHandler` is the right interface for depending on the datapath "nodehandler". It's up to the package `pkg/node/manager` to define its own version `manager.NodeHandler` that is used for subscribing to the `NodeManager`.

joamaki commented on pkg/datapath/cells.go:
```diff
@@ -217,8 +217,6 @@ type datapathParams struct {
 
 	BandwidthManager types.BandwidthManager
 
-	ModulesManager *modules.Manager
```
Consider squashing this with the previous commits that removed fields from here.

mhofstetter replied:
done (squashed all commits that are removing unnecessary deps from the `datapathParams`)

joamaki reviewed (approved):
:tada:

mhofstetter committed:
datapath: move BpfMaps dependencies to loader cell

Currently, the datapath cell depends on all BPF maps that are
provided as Hive Cell. The initial reason is that the loader
initialization starts when all BPF maps are registered.

But with the extraction of the loader into its own Hive Cell, the loader
Hive Cell should actually be the one depending on the BPF maps.

Therefore, this commit moves the dependencies from the datapath to the
loader hive cell.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>
